### DISCUSSION TITLE
Help with testing a sequence of two commands using PHPUnit and ReactPHP

### DISCUSSION THREAD

**Discussion by ghost** [reactphp/reactphp#537] [Q&A]
👋 Hi there,

At ChesslaBlab we're using PHPUnit along with `React\Socket\Connector` to write [functional tests](https://github.com/chesslablab/chess-server/tree/master/tests/functional) for the PHP Chess Server. 

The following example shows how a sequence of one command can be tested successfully.

```php
    // ...

    /**
     * @test
     */
    public function start_classical_fen()
    {
        self::$connector->connect("$this->host:$this->port")->then(function (ConnectionInterface $conn) {
            $conn->on('data', function ($data) use ($conn) {
                self::$deferred->resolve($data);
                $conn->close();
            });
            $conn->write("/start classical fen");
        });

        $response = \React\Async\await(self::$promise->then(function (string $result): string {
            return $result;
        }));

        $expected = '{"\/start":{"variant":"classical","mode":"fen","fen":"rnbqkbnr\/pppppppp\/8\/8\/8\/8\/PPPPPPPP\/RNBQKBNR w KQkq -"}}';

        $this->assertEquals($expected, $response);
    }

    // ...
}
```

However, it seems as if we're wissing something when testing a sequence of two commands.

```
/start classical fen
/legal e2
```

In this case, the expected response should be the following.

```php
$expected = '{"\/legal":{"color":"w","id":"P","fen":{"e3":"rnbqkbnr\/pppppppp\/8\/8\/8\/4P3\/PPPP1PPP\/RNBQKBNR b KQkq -","e4":"rnbqkbnr\/pppppppp\/8\/8\/4P3\/8\/PPPP1PPP\/RNBQKBNR b KQkq e3"}}}';
```

See:

- https://github.com/chesslablab/chess-server/issues/190

I haven't been able to find an example of how to test a sequence of two commands using PHPUnit and ReactPHP. 

Any help would be greatly appreciated, thank you!

**Comment by kaifkh20** [comment]
I think presently there is now way to do this I think.

But I have found an another way that reuse the same function from the StartCommandTest class. Would you like to see it??

**Comment by ghost** [comment]
See https://github.com/orgs/chesslablab/discussions/381

**Comment by WyriHaximus** [comment]
I'm somewhat confused about why the two commands would fail. Showed code should handle that fine? 

You could rewrite:

```php
        $response = \React\Async\await(self::$promise->then(function (string $result): string {
            return $result;
        }));
```

As:

```php
        $response = \React\Async\await(self::$promise);
```

**Comment by kaifkh20** [comment]
I don't think that would work since conn does not let us write two functions at a time.
I think the problem is when we write the commands in conn->write function it only let us write only once.

**Reply by WyriHaximus** [reply]
conn->write lets you write anything you want. It's a raw TCP stream so you'd have to build or implement a protocol on top of that to handle commands.

**Comment by ghost** [comment]
Thank you @WyriHaximus for the help.

Not too sure about this one yet. No worries, I suppose for the time being the Selenium automated tests will be kind of equivalent to the so-called functional tests.
